src/internal/Observable.ts

```diff
@@ -392,4 +392,123 @@ export class Observable<T> implements Subscribable<T> {
   pipe(...operations: OperatorFunction<any, any>[]): Observable<any> {
     return pipeFromArray(operations)(this);
   }
+
+  /**
+   * Observable is async iterable, so it can be used in `for await` loop. This method
+   * of subscription is cancellable by breaking the for await loop. Although it's not
+   * recommended to use Observable's AsyncIterable contract outside of `for await`, if
+   * you're consuming the Observable as an AsyncIterable, and you're _not_ using `for await`,
+   * you can use the `throw` or `return` methods on the `AsyncGenerator` we return to
+   * cancel the subscription. Note that the subscription to the observable does not start
+   * until the first value is requested from the AsyncIterable.
+   *
+   * Functionally, this is equivalent to using a {@link concatMap} with an `async` function.
+   * That means that while the body of the `for await` loop is executing, any values that arrive
+   * from the observable source will be queued up, so they can be processed by the `for await`
+   * loop in order. So, like {@link concatMap} it's important to understand the speed your
+   * source emits at, and the speed of the body of your `for await` loop.
+   *
+   * ## Example
+   *
+   * ```ts
+   * import { interval } from 'rxjs';
+   *
+   * async function main() {
+   *  // Subscribe to the observable using for await.
+   *  for await (const value of interval(1000)) {
+   *    console.log(value);
+   *
+   *    if (value > 5) {
+   *      // Unsubscribe from the interval if we get a value greater than 5
+   *      break;
+   *    }
+   *  }
+   * }
+   *
+   * main();
+   * ```
+   */
+  [Symbol.asyncIterator](): AsyncGenerator<T, void, void> {
+    let subscription: Subscription | undefined;
+    let hasError = false;
+    let error: any;
+    let completed = false;
+    const values: T[] = [];
+    const deferreds: [(value: any) => void, (reason: any) => void][] = [];
+
+    const handleError = (err: any) => {
+      hasError = true;
+      error = err;
+      while (deferreds.length) {
+        deferreds.shift()![1](err);
+      }
+    };
+
+    const handleComplete = () => {
+      completed = true;
+      while (deferreds.length) {
+        deferreds.shift()![0]({ value: undefined, done: true });
+      }
+    };
+
+    return {
```

Comment:
You may (or may not) want to make this object inherit from `AsyncIterator.prototype`, which already exists but in [the async-iterator-helpers proposal](https://github.com/tc39/proposal-async-iterator-helpers) will gain additional helper methods like `map`. (Prior to that proposal shipping, the only thing inheriting gives you is the implementation of `Symbol.asyncIterator` you have below.)

`AsyncIterator` isn't currently a global, but you can today access `AsyncIterator.prototype` by doing `(async function*(){})().__proto__.__proto__.__proto__`.

---

That said, in that proposal the new `AsyncIterator.from` helper is the intended way for users to get async iterators out of things, and that will wrap the result up in a wrapper which inherits from `AsyncIterator.prototype`, so `AsyncIterator.from(observable)` will have the new helper methods on it either way. So even in the current state the only way users could get at a thing without the new helpers would be by doing `observable[Symbol.asyncIterator]()`. Still, that's a thing they might do, so I think it would make sense to inherit.

Reply:
We don't want to do that. This an interop point, I don't want people to deoptimize a pure push type like Observable to a pull/push type like AsyncIterable just to `map`. We have a `map` we export.